src/librustc/ty/sty.rs

```diff
@@ -2407,18 +2407,91 @@ pub struct Const<'tcx> {
 #[cfg(target_arch = "x86_64")]
 static_assert_size!(Const<'_>, 48);
 
+/// Returns the `DefId` of the constant parameter that the provided expression is a path to.
+fn const_param_def_id(expr: &hir::Expr<'_>) -> Option<DefId> {
+    // Unwrap a block, so that e.g. `{ P }` is recognised as a parameter. Const arguments
+    // currently have to be wrapped in curly brackets, so it's necessary to special-case.
+    let expr = match &expr.kind {
+        hir::ExprKind::Block(block, _) if block.stmts.is_empty() && block.expr.is_some() => {
+            block.expr.as_ref().unwrap()
+        }
+        _ => expr,
+    };
+
+    match &expr.kind {
+        hir::ExprKind::Path(hir::QPath::Resolved(_, path)) => match path.res {
+            hir::def::Res::Def(hir::def::DefKind::ConstParam, did) => Some(did),
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
 impl<'tcx> Const<'tcx> {
+    /// Literals and const generic parameters are eagerly converted to a constant, everything else
+    /// becomes `Unevaluated`.
+    pub fn from_hir_anon_const(
+        tcx: TyCtxt<'tcx>,
+        ast_const: &hir::AnonConst,
+        ty: Ty<'tcx>,
+    ) -> &'tcx Self {
```

Comment:
IMO this should take `DefId` and `Substs` and create `Unevaluated` by default.

The `hir::AnonConst` special-casing is just a hack to make a few things work better.

Reply:
that makes the function less useful than it is now. It abstracts away the HirId -> DefId conversion and substs creation, which would need to get replicated at all call sites. I can split it into two functions if you prefer, but I'd like to keep the current API

Reply:
> It abstracts away the HirId -> DefId conversion and substs creation, which would need to get replicated at all call sites

`DefId` is the correct type here, it's just unfortunate that the HIR contains a `HirId` instead. And substs are easy to get.

And if you have more than 2-3 calls to this function something went wrong.